[TG-7648] Constant propagation of substring operations

danpoe opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Implement constant propagation for substring operations

This implements constant propagation of ID_cprover_string_substring_func
function application expressions. These are in turn used by the various
substring operations of String and StringBuilder.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eaeadb62f60b7dce81bd2c69fcc194f742d690ab).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120411659

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 415535ecd286be4ffac801726a11676251f51ea9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122582457
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

smowton commented on jbmc/regression/jbmc-strings/StringBuilderSubstringConstantEvaluation1/Main.java:
```diff
@@ -0,0 +1,25 @@
+    String s5 = sb.substring(0, 1);
+    assert s5.length() == 1;
+    assert s5.startsWith("a");
+  }
```
Additionally check that out-of-range begin- or end-indices still throw as expected?

allredj replied:
@smowton Exceptions are a thing of the model library. There are already a couple of tests there:
https://github.com/diffblue/models-library/blob/develop/modelTests/maven/java/lang/String/src/main/java/L1Substring_int.java
https://github.com/diffblue/models-library/blob/develop/modelTests/maven/java/lang/String/src/main/java/L1Substring_int_int.java
which seem sufficient to me.

Of course these would only fail when the test-gen bump is updated, but I don't see how they coiuld.

danpoe replied:
I think @smowton is referring to the case where all the arguments to `substring()` are constant, but the indices are out of bounds. In this case the constant propagator should not simplify anything and the statement should just be symex'd normally. Whether that then later leads to an exception is a thing of the models library. I'll add a test to check that no simplification occurs for out of bounds indices.

smowton commented on jbmc/regression/jbmc-strings/StringSubstringConstantEvaluation2/test3.desc:
```diff
@@ -0,0 +1,9 @@
+CORE symex-driven-lazy-loading-expected-failure
```
Any idea why the xfail?

danpoe replied:
When using symex driven lazy loading it seems that jbmc does not fully honor the `--property` option as several additional VCCs are generated (e.g. null pointer checks). Those regression tests though check that a specific VCC is not simplified. This is currently done by selecting that VCC via `--property` and then checking that jbmc prints `Generated 1 VCC(s), 1` remaining after simplification. If there would be more VCCs we could not check that this one VCC is the one that is not simplified.

smowton commented on src/goto-symex/goto_symex.cpp:
```diff
@@ -310,6 +314,26 @@ goto_symext::try_evaluate_constant_string(
   return try_get_string_data_array(s_pointer_opt->get(), ns);
 }
 
+optionalt<std::reference_wrapper<const constant_exprt>>
```
`optionalt<reference_wrapper<T>>` -> `T*`

danpoe replied:
It was suggested in a previous PR to use optional references as they provide some additional semantic information. I don't really mind either way but we should be consistent about it.

romainbrenguier replied:
I don't think there is a semantic difference, a reference wrapper is basically a pointer that can't be null, by making it optional you can get the empty optional which is basically equivalent to a null pointer, so it's just a verbose way of saying the same thing.

danpoe replied:
Yes, I agree that they behave the same. What I meant was that when using an `optionalt` it's clear from the signature of the method that it returns an optional value, whereas when using a plain pointer that's not necessarily clear.

smowton replied:
I suppose I would always assume a pointer might be null, otherwise you'd use a reference

danpoe replied:
Sure, though a nullpointer return value could suggest that some kind of error has occured, whereas `optionalt` would suggest that this is part of normal and expected behavior.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e0ef78e57eafa62c9f0709f02ba31de29401412d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123016840

danpoe committed:
Add String.substring() and StringBuilder.substring() regression error tests

Add regression tests that check that the constant propagator does not simplify
substring() invocations with only constant arguments when the indices are out of
bounds.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7da4f78b6e5edb66afb27ec9cb9f17ba6ad179aa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123397270
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e56041f4cf250f77a010b7231c74a3fa6051b1a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123399979
